torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -141,25 +142,32 @@ void ProcessGroupAgent::enqueueSend(SendWork work) {
     [&](const SendWork& work) {
       std::stringstream ss;
       serialize(work.message_, ss);
-      std::string str = ss.str();
+      std::string serializedPayload = ss.str();
 
       std::vector<torch::Tensor> preamble = {
         torch::tensor(
           {
             (int64_t)pg_->getRank(),
-            (int64_t)str.length(),
+            (int64_t)serializedPayload.length(),
             (int64_t)work.message_.type()
           }, {torch::kLong})
       };
 
       // ProcessGroup is not thread-safe when sending with the same tag, hence
       // the lock
-      std::lock_guard<std::mutex> lock(sendMutexes_[work.to_]);
-      pg_->send(preamble, work.to_, work.to_ /* channelTag */)->wait();
-      if (work.message_.type() != MessageType::SHUTDOWN) {
-        std::vector<torch::Tensor> payload ={
-            torch::from_blob((void *)str.c_str(), str.length(), {torch::kChar})
+      if (work.message_.isShutdown()) {
+        std::lock_guard<std::mutex> lock(sendMutexes_[work.to_]);
+        pg_->send(preamble, work.to_, work.to_ /* channelTag */)->wait();
+      } else {
+        std::vector<torch::Tensor> payload = {
+            torch::from_blob(
+                (void *)serializedPayload.c_str(),
+                serializedPayload.length(),
+                {torch::kChar}
+            )
         };
+        std::lock_guard<std::mutex> lock(sendMutexes_[work.to_]);
+        pg_->send(preamble, work.to_, work.to_ /* channelTag */)->wait();
```

Comment:
As I mentioned in my previous comment, I don't think we need to wait under the lock here.